pyfolio/perf_attrib.py

```diff
@@ -0,0 +1,138 @@
+
+    positions: pd.DataFrame
+        Daily holdings (in dollars or percentages), indexed by date.
+        Will be converted to percentages if positions are in dollars.
+        Short positions show up as cash in the 'cash' column.
+        - Examples:
+                        AAPL  TLT  XOM  cash
+            2017-01-01    34   58   10     0
+            2017-01-02    22   77   18     0
+            2017-01-03   -15   27   30    15
+
+                            AAPL       TLT       XOM  cash
+            2017-01-01  0.333333  0.568627  0.098039   0.0
+            2017-01-02  0.188034  0.658120  0.153846   0.0
+            2017-01-03  0.208333  0.375000  0.416667   0.0
+
+    factor_returns : pd.DataFrame
+        Returns by factor, with date as index and factors as columns
+        - Example:
+                        momentum  reversal
+            2017-01-01  0.002779 -0.005453
+            2017-01-02  0.001096  0.010290
+
+    factor_loadings : pd.DataFrame
+        Factor loadings for all days in the date range, with date and ticker as
+        index, and factors as columns.
+        - Example:
+                               momentum  reversal
+            dt         ticker
+            2017-01-01 AAPL   -1.592914  0.852830
+                       TLT     0.184864  0.895534
+                       XOM     0.993160  1.149353
+            2017-01-02 AAPL   -0.140009 -0.524952
+                       TLT    -1.066978  0.185435
+                       XOM    -1.798401  0.761549
+
+    Returns
+    -------
+    perf_attribution : pd.DataFrame
+        df with factors, common returns, and specific returns as columns,
+        and datetimes as index
+        - Example:
+                        momentum  reversal  common_returns  specific_returns
+            dt
+            2017-01-01  0.249087  0.935925        1.185012          1.185012
+            2017-01-02 -0.003194 -0.400786       -0.403980         -0.403980
+    """
+    # convert holdings to percentages, and convert positions to long format
+    positions = get_percent_alloc(positions)
+    # remove cash after normalizing positions
+    del positions['cash']
+    positions = positions.stack()
+    positions.index = positions.index.set_names(['dt', 'ticker'])
+
+    risk_exposures = factor_loadings.multiply(positions,
+                                              axis='rows')
+
+    risk_exposures_portfolio = risk_exposures.groupby(level='dt').sum()
+    perf_attrib_by_factor = risk_exposures_portfolio.multiply(factor_returns)
+
+    common_returns = perf_attrib_by_factor.sum(axis='columns')
+    specific_returns = returns - common_returns
+
+    returns_df = pd.DataFrame({'total_returns': returns,
+                               'common_returns': common_returns,
+                               'specific_returns': specific_returns})
+
+    return pd.concat([perf_attrib_by_factor, returns_df], axis='columns')
+
+
+def create_perf_attrib_stats(perf_attrib):
```

Comment:
I don't think so, it just calls a few functions that all have unittests in their respective packages.